x/executionlayer/client/cli/tx.go

```diff
@@ -20,6 +20,114 @@ import (
 	"github.com/spf13/viper"
 )
 
+func GetCmdContractRun(cdc *codec.Codec) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "run <type> <wasm-path>|<uref>|<name>|<hash> <argument> <fee> <gas_price> --wallet|--address|--nickname <from>",
+		Short: "Run contract",
+		Long: "Run contract\n" +
+			"There are 4 types of contract run. ('wasm', 'uref', 'name', 'hash)",
+		Args: cobra.ExactArgs(5),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			txBldr := auth.NewTxBuilderFromCLI().WithTxEncoder(utils.GetTxEncoder(cdc))
+			cliCtx := context.NewCLIContext().WithCodec(cdc)
+
+			// Extract "from" from flags
+			var fromAddr sdk.AccAddress
+
+			kb, err := client.NewKeyBaseFromDir(viper.GetString(client.FlagHome))
+			if err != nil {
+				return err
+			}
+
+			if walletname := viper.GetString(FlagWallet); walletname != "" {
+				key, err := kb.Get(walletname)
+				if err != nil {
+					return err
+				}
+
+				fromAddr = key.GetAddress()
+				cliCtx = cliCtx.WithFromAddress(fromAddr).WithFromName(key.GetName())
+			} else if straddr := viper.GetString(FlagAddress); straddr != "" {
+				fromAddr, err = sdk.AccAddressFromBech32(straddr)
+				if err != nil {
+					return fmt.Errorf("malformed address in --address: %s\n%s", straddr, err.Error())
+				}
+
+				key, err := kb.GetByAddress(fromAddr)
+				if err != nil {
+					return err
+				}
+
+				cliCtx = cliCtx.WithFromAddress(fromAddr).WithFromName(key.GetName())
+			} else if nickname := viper.GetString(FlagNickname); nickname != "" {
+				fromAddr, err = cliutil.GetAddress(cliCtx.Codec, cliCtx, nickname)
+				if err != nil {
+					return fmt.Errorf("no registered address of the given nickname '%s'", nickname)
+				}
+
+				key, err := kb.GetByAddress(fromAddr)
+				if err != nil {
+					return err
+				}
+				cliCtx = cliCtx.WithFromAddress(fromAddr).WithFromName(key.GetName())
+			} else {
+				return fmt.Errorf("one of --address, --wallet, --nickname is essential")
+			}
+
+			sessionType := cliutil.GetContractType(args[0])
+			var sessionCode []byte
+			switch sessionType {
+			case util.WASM:
+				sessionCode = util.LoadWasmFile(args[1])
+			case util.HASH:
+			case util.UREF:
+				sessionCode = util.DecodeHexString(args[1])
+			case util.NAME:
+				sessionCode = []byte(args[1])
+			default:
+				return fmt.Errorf("type must be one of wasm, name, uref, or hash")
+			}
+
+			sessionArgs, err := cliutil.ProtobufSafeDecodeingHexString(args[2])
```

Comment:
Later, we should be think about how to deal with users' various custom WASMs.

Before sending to EE, each args should be converted into byte, but we don't know the type in this state.

3 choices to implement that:

1. User should convert into byte, serialize, and hexilify.
1. User compiles twice. One for WASM, one for JSON or YAML containing functions and types of args list, and put into system in `define` stage. Then user puts parameter as defined in rust contract.
1. User puts parameter with type flags such as `1:uint256,"bryan":str,9999:uint8` then, system can convert & serialize into hexilified ABI form.

We may discuss next week.

Reply:
Currently, hex string(must be casperlabs abi) is inputted and converted. Let's consider some more formats such as json and reflect them further.

Reply:
There are some formats in Supported types of contract arguments of https://github.com/CasperLabs/CasperLabs/blob/dev/docs/CONTRACTS.md.

Reply:
Then we practically cannot deploy a contract with args at present. right~?

Reply:
No, as stated above, you can send abi as hex.

Reply:
@JoowonYun I am just curious that how can I convert `[{"name": "target_address", "value": {"bytes_value": "6e88ed546646f6613224351c1695bd583dad138999b397bffa4d541fe99ebc59"}}, {"name": "amount", "value": {"long_value": 100000000}}]` to hex?

Reply:
`name` is not important. `str` is important.

If WASM function is `blahblah(str&: target_address, str& name, u256 amount)`, the args will come in `str, str, u256` order. (Serialize) Then, each args will be converted by `str to byte(or hex)`, `str to byte(or hex)`, `uint to byte(or hex)` and combined together. The combined will be send to protobuf.

So, the reason why JSON or YAML should store is, the order of args and these types.

Reply:
@psy2848048 Have you read about the `Supported types of contract arguments` in this [doc](https://github.com/CasperLabs/CasperLabs/blob/dev/docs/CONTRACTS.md)?

There are already used JSON format for contract's argument following [proto3 JSON format](https://developers.google.com/protocol-buffers/docs/proto3#json)

The reason why I mentioned about "practical feasibility" is to check the current state of the passing argument feature. Because It seems there are no ways to deploy a contract with args from a cli.(A user should generate abi byte sequence.)

Reply:
According to the implementation gRPC util..
https://github.com/hdac-io/casperlabs-ee-grpc-go-util/blob/master/util/util.go#L116
CasperEE just receives serialized bytes and send as Args.

The JSON form which you said is implemented as like this:
https://github.com/hdac-io/casperlabs-ee-grpc-go-util/blob/master/protobuf/io/casperlabs/ipc/ipc.pb.go#L29